lib/escaping.js

```diff
@@ -0,0 +1,43 @@
+function escape(propertyName) {
+  return propertyName.replace(/\\/g, '\\\\').replace(/\./g, '\\.');
+}
+
+function unescapeAndSplit(path) {
+  const properties = [];
+  let i = path.indexOf('.');
+  while (i >= 0) {
+    if (isEscapedDot(path, i)) {
+      path = unescapeFirstDotChar(path);
```

Comment:
Can we avoid reassigning the argument value and introduce a new constant instead?

Reply:
The core logic of this function is based on the mutability of the `path` variable (as we unescape the escaped dots in place, then cut leading property names and handle the rest of the path). Re-writing it in an "immutable" way is possible but IMO more complex and not sure it's worth it.

If you just don't like the idea of mutating parameters in general then maybe this is what you mean?
```
function unescapeAndSplit(originalPath) {
   let path = originalPath;
   ... // rest of the code as it is now
}
```

If so - sure, I can change it.

Reply:
yeah, that works!

Reply:
Done